### PR TITLE
Feature/1674 column meaning

### LINKED ISSUE #1674: Make ColumnMeaning an interface and allow its implementations
ColumnMeaning is now a simple class containing already so many values. For specific sub-projects, new specific values are required. Extension of this class made it already too large for simple use (drop-down box list is too long).

Let's make ColumnMeaning an interface and create specific implementations of this interface.

And let's make the drop-down more readable/user-friendly:
drop-down

### PR DESCRIPTION
**Pull request by jhorcicka**
Resolves #1674 

Hard-coded ColumnMeaning feature(s) for input columns mapping was refactored so that every wizard can provide its own HasColumnMeaning (interface) implementation. This should make combo-box content more usable leaving unneeded general items out. The combo-box content was also made group-aware. 

group-combo-box

### CODE REVIEW

File: api/src/main/java/org/datacleaner/metadata/DefaultColumnMeaningCollection.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        final HasColumnMeaning[] values = ColumnMeaning.class.getEnumConstants();
+
+        for (final HasColumnMeaning columnMeaning : values) {
+            populateMatchMap(columnMeaning.getName(), columnMeaning);
+
+            for (final String alias : columnMeaning.getAliases()) {
+                populateMatchMap(alias, columnMeaning);
+            }
+        }
+    }
```
Comment (LosD) [inline api/src/main/java/org/datacleaner/metadata/DefaultColumnMeaningCollection.java]: Such a big static initialiser is a little bit nasty. Can we do this lazily instead?

**Comment by LosD** [comment]
Code LGTM, I'll do a quick test

**Comment by LosD** [comment]
During testing, I saw one thing: There's no grouping in the default column meaning implementation, but there was still what looks like some weird "Meanings" group:
image

**Comment by LosD** [comment]
If possible, when there _are_ groups, I think it would look nice if the non-group elements are indented few pixels (I'm basing this on my assumption that the "Meanings" above indeed is a group header.

If there isn't any groups, it's probably better to keep the meanings themselves as top-level

**Comment by jhorcicka** [comment]
Yes, *Meanings* is a default null-group-name. I'll skip that and provide indentation if there are real groups.

**Comment by jhorcicka** [comment]
@LosD: updated, ready for re-review.

File: desktop/ui/src/main/java/org/datacleaner/widgets/DCGroupComboBox.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+ * Free Software Foundation, Inc.
+ * 51 Franklin Street, Fifth Floor
+ * Boston, MA  02110-1301  USA
+ */
+package org.datacleaner.widgets;
+
+import java.awt.Font;
+
+import javax.swing.DefaultComboBoxModel;
+import javax.swing.DefaultListCellRenderer;
```
Comment (arjansh) [inline desktop/ui/src/main/java/org/datacleaner/widgets/DCGroupComboBox.java]: The import javax.swing.DefaultListCellRenderer is never used

File: desktop/ui/src/main/java/org/datacleaner/widgets/DCGroupComboBox.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+import javax.swing.JLabel;
+import javax.swing.JList;
+
+import org.apache.metamodel.util.HasName;
+import org.datacleaner.util.StringUtils;
+
+/**
+ * Defines a group combo-box class that has a more convenient listening mechanism.
+ */
+public class DCGroupComboBox extends DCComboBox<HasName> {
```
Comment (arjansh) [inline desktop/ui/src/main/java/org/datacleaner/widgets/DCGroupComboBox.java]: The serializable class DCGroupComboBox does not declare a static final serialVersionUID field of type long

File: desktop/ui/src/main/java/org/datacleaner/widgets/DCGroupComboBox.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+        @Override
+        public String getName() {
+            return _text;
+        }
+    }
+
+    public DCGroupComboBox() {
+        super(new GroupComboBoxModel());
```
Comment (arjansh) [inline desktop/ui/src/main/java/org/datacleaner/widgets/DCGroupComboBox.java]: Type safety: The expression of type DCGroupComboBox.GroupComboBoxModel needs unchecked conversion to conform to ComboBoxModel<HasName>

File: desktop/ui/src/main/java/org/datacleaner/widgets/DCGroupComboBox.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        setRenderer(new GroupComboBoxRenderer());
+    }
+
+    public void addDelimiter(String text) {
+        _groupExists = true;
+        addItem(new Delimiter(text));
+    }
+
+    private static class GroupComboBoxModel extends DefaultComboBoxModel {
```
Comment (arjansh) [inline desktop/ui/src/main/java/org/datacleaner/widgets/DCGroupComboBox.java]: The serializable class GroupComboBoxModel does not declare a static final serialVersionUID field of type long.
DefaultComboBoxModel is a raw type. References to generic type DefaultComboBoxModel<E> should be parameterized.

File: desktop/ui/src/main/java/org/datacleaner/widgets/DCGroupComboBox.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                final int index = getIndexOf(item);
+
+                if (index < getSize()) {
+                    setSelectedItem(getElementAt(index + 1));
+                }
+            }
+        }
+    }
+
+    private class GroupComboBoxRenderer extends EnumComboBoxListRenderer {
```
Comment (arjansh) [inline desktop/ui/src/main/java/org/datacleaner/widgets/DCGroupComboBox.java]: The serializable class GroupComboBoxRenderer does not declare a static final serialVersionUID field of type long

File: desktop/ui/src/main/java/org/datacleaner/widgets/DCGroupComboBox.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        @Override
+        public JLabel getListCellRendererComponent(final JList list, final Object value, final int index,
```
Comment (arjansh) [inline desktop/ui/src/main/java/org/datacleaner/widgets/DCGroupComboBox.java]: JList is a raw type. References to generic type JList<E> should be parameterized

**Comment by jhorcicka** [comment]
@arjansh: updated, ready for re-review.

**Pull request merged**